alpenhorn/cli/file/list.py

```diff
@@ -0,0 +1,382 @@
+    options to list places to search.
+
+    The first two of these negative constraints (nodes and/or groups which must not
+    have the files) and the second two are positive constraints (nodes and/or groups
+    which must have the file).  For a file to be listed, it must satisfy both the
+    negative and positive constraints.
+
+    Normally, only one of the location constraints given need be satisfied for a
+    file to be listed, but if you also use "--all", then a file will be listed only
+    if all location constraints are satisfied.
+
+    \b
+    Limit by file state
+    -------------------
+
+    To limit the list based on file state, you can use the state flags:
+    --missing, --corrupt, --suspect, --healthy.  In this case, you must specify
+    at least one location (with --node or --group) to indicate where the file
+    state should be checked.  If multiple state flags are used, files with any
+    one of the specified state will be listed.
+
+    Note: the "missing" state does not mean simply "absent".  The state "missing"
+    is a special state tracked by alpenhorn in the data index for files which are
+    absent from a node/group when they are expected to be present (i.e they haven't
+    been released).  So, these two commands produce (potentially) different results:
+
+    \b
+    group list --absent-node=NODE
+    group list --missing --node=NODE
+
+    The first lists all files absent from NODE.  The second lists only files which are
+    absent but not released from the NODE (i.e. files which have been deleted by a
+    third-party without alpenhorn's knowledge).
+
+    \b
+    Limit based on syncability
+    --------------------------
+
+    Using both "--from=SRC" and "--to=DEST" will limit the list to files which
+    can be synced (copied) from the Node SRC to the Group DEST.  Specifically,
+    this restricts the list to files which are healthy on SRC and absent from
+    DEST.
+
+    You cannot simultaneously limit both by file state and syncability.
+    """
+
+    # Usage checks.
+
+    # Find a state flag to complain about, if any
+    if corrupt:
+        state_flag = "corrupt"
+    elif missing:
+        state_flag = "missing"
+    elif healthy:
+        state_flag = "healthy"
+    elif suspect:
+        state_flag = "suspect"
+    else:
+        state_flag = None
+
+    # Must use --to and --from together
+    both_or_neither(from_, "from", to, "to")
+    # Can't use a state with --from or --to
+    # (checking --from is enough, given the previous test)
+    not_both(from_, "from", state_flag, state_flag)
+
+    # Can't use --all if we haven't used a location constraint
+    if all_:
+        if not node and not group and not absent_group and not absent_node:
+            raise click.UsageError(
+                "--all cannot be used without a location constraint "
+                "(one or more of --absent-group, --absent-node, --group, --node)"
+            )
```

Comment:
Hmm, true.  Using `--all` with a single constraint isn't going to cause trouble, but you're right there needs to be more than one for it to have any effect.

Really, there's no problem with using `--all` with _no_ location constraint.  It would just be ignored.  The usage check here is mostly to re-inforce that `--all` applies to location constraints only.

I'm not sure it's worth counting how many location constraints have been used.